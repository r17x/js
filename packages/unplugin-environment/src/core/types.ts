import type { ZodObject, ZodRawShape } from "zod";
/**
 * @description Match option defines the prefixes for the environment variables that will be loaded. You can pass a single string, an array of strings for multiple prefixes, or exclude variables using a ! before the prefix.
 * @example
 * ```ts
 * Environment("PREFIX_APP")
 * Environment(["PREFIX_APP", "PREFIX2"])
 * Environment(["PREFIX_APP", "PREFIX2", "!PREFIX3"])
 * ````
 */
export type Match = string | string[];

type ClientServer<T> = { client: T; server: T };
/**
 * @description Options with schema and will "strict" validate the env variable
 * @example
 * ```ts
 * Environment({
 *   match: 'REACT_APP',
 *   schema: { APP_NAME: z.string() }
 * })
 *
 * Environment({
 *   match: 'REACT_APP',
 *   schema: { APP_NAME: z.string() },
 *   moduleEnvName: '@myenv'
 * })
 * ```
 */
export type Options = {
	match: Match;
	schema: ZodObject<ZodRawShape>;
	moduleEnvName: string;
};

export type OptionsClientServer = ClientServer<Options>;
export type UOptions = Options | OptionsClientServer;

type SchemaOption = {
	match: Match;
	schema: ZodObject<ZodRawShape> | ZodRawShape;
	/**
	 * @default '@env'
	 */
	moduleEnvName?: string;
};

type SchemaClientServerOption = ClientServer<Match | SchemaOption>;

/**
 * @example
 * ```ts
 * // When you need only one prefix.
 * Environment("PREFIX_APP")
 * // When you need multiple prefixes.
 * Environment(["PREFIX_APP", "PREFIX2"])
 * // When you need validation for the env variables.
 * Environment({
 *   match: 'PREFIX_APP',
 *   schema: { PREFIX_APP_NAME: z.string() }
 * })
 * // When you want to customize the module name for accessing the env variables.
 * Environment({
 *   match: 'REACT_APP',
 *   schema: { PREFIX_APP_NAME: z.string() },
 *   moduleEnvName: '@myenv'
 * })
 * // When you need different environment for client and server.
 * Environment({
 *  client: {
 *   match: 'CLIENT_APP',
 *   schema: { CLIENT_APP_NAME: z.string() },
 *   moduleEnvName: '@myenv/client'
 *  },
 *  server: {
 *   match: 'SERVER_APP',
 *   schema: { DB_URL: z.url() },
 *   moduleEnvName: '@myenv/server'
 *  }*
 * })
 * ```
 */
export type PluginOption = Match | SchemaOption | SchemaClientServerOption;
