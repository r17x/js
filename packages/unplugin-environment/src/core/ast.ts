// AST Helpers for generated typescript definition
// this module will be used fro generated /types/env.d.ts based on user schema.
import * as ts from "typescript";

const printer = ts.createPrinter({
	newLine: ts.NewLineKind.LineFeed,
	omitTrailingSemicolon: true,
});

const emptySourceFile = ts.createSourceFile("", "", ts.ScriptTarget.Latest);

export const printTypeDefinition = (node: ts.TypeNode) =>
	printer.printNode(ts.EmitHint.Unspecified, node, emptySourceFile);

export { SyntaxKind } from "typescript";
export { zodToTs } from "zod-to-ts";
