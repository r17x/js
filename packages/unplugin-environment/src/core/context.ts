import type { Options } from "@/types";

class Context {
	options: Options;
	env: unknown;

	constructor(options: Options) {
		this.options = options;
	}

	getNewId(_id: string) {}

	resolveId(_id: string, _importer?: string) {}

	filter(_id: string) {}

	log(_kind: "error" | "info", _message: string) {}

	async build() {}

	watchChange(_id: string, _change: string) {}
}

export { Context };
