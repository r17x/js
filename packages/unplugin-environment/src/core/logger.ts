import colors from "picocolors";

export function log(
	{
		kind,
		message,
		prefix: p,
	}: { kind: "error" | "info"; message: string; prefix?: string },
	...args: unknown[]
) {
	const timeFormatter = new Intl.DateTimeFormat(undefined, {
		hour: "numeric",
		minute: "numeric",
		second: "numeric",
	});

	const prefix = [
		colors.dim(timeFormatter.format(new Date())),
		kind === "error"
			? colors.dim(colors.red(`[${[p, colors.bold("failed")].join(":")}]`))
			: colors.cyan(`[${[p, colors.bold("success")].join(":")}]`),
	].join(" ");

	message.split("\n").forEach((line, index, arr) => {
		line.length <= 2
			? console[kind](prefix.replaceAll(/.*/g, " "), line)
			: kind === "error" &&
					(line.includes("│") || arr[index - 2]?.includes("│"))
				? console[kind](prefix, colors.bold(line))
				: console[kind](prefix, line);
	});
	if (args.length > 0) {
		console.log("\n", JSON.stringify(args, null, 2), "\n");
	}
}
