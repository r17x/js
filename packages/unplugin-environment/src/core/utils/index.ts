import { createFilter } from "@rollup/pluginutils";

const filter = createFilter([/\.res$/], [/lib\/bs\//]);

export { filter };
export * from "./load-resconfig";
