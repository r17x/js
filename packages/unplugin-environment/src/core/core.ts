import path from "path";
import { A, D, F, G, S, pipe } from "@mobily/ts-belt";
import { z } from "zod";
import { SyntaxKind, printTypeDefinition, zodToTs } from "./ast";
import type { Options, PluginOption } from "./types";

const toUndefined = () => undefined;
const exclaim = S.startsWith("!");
const dropHead = S.sliceToEnd(1);

const matchSchema = z.string().min(1).or(z.string().min(1).array());

const defaultOption = (match: Options["match"]) => ({
	match,
	schema: z.record(z.string().optional()),
	moduleEnvName: "@env",
});

const strictOption = z.strictObject({
	match: matchSchema,
	schema: z
		.record(z.instanceof(z.ZodType))
		.transform((a) => z.object(a))
		.or(z.instanceof(z.ZodObject)),
	moduleEnvName: z.string().optional().default("@env"),
});

const option = matchSchema.transform(defaultOption).or(strictOption);

export const getOptions = (o: PluginOption): Options =>
	option.parse(o) as Options; // using as for type but don't worries, we have power of `zod`.

const filterEnv = (env: Record<string, string>, m: Options["match"]) => {
	const ms = G.isString(m) ? [m] : m;
	const white = pipe(ms, A.filter(G.isNot(exclaim)));
	const black = pipe(ms, A.filter(exclaim), A.map(dropHead));
	const selector = (key: string) => A.some(white, (m) => S.startsWith(key, m));

	return pipe(env, D.filterWithKey(selector), D.deleteKeys(black));
};

export const getEnv = (env: Record<string, string>, options: Options) => {
	const filteredEnv = filterEnv(env, options.match);

	return F.ifElse(
		options.schema.safeParse(filteredEnv),
		({ success }) => success,
		({ data }) => D.merge(filteredEnv, data),
		() => filteredEnv,
	);
};

export const createModuleEnv = (
	env: Record<string, string>,
	options: Options,
) => ({
	code: `export const env = ${JSON.stringify(getEnv(env, options))}`,
});

export const createModuleDTS = (
	env: Record<string, string>,
	options: Options,
) =>
	pipe(
		D.keys(filterEnv(env, options.match)),
		A.reduce(D.makeEmpty<Record<string, z.ZodType>>(), (a, b) =>
			D.set(a, b, z.string().optional()),
		),
		z.object,
		(zo) => zo.merge(options.schema),
		zodToTs,
		({ node }) => node,
		printTypeDefinition,
		(typeDef) => [
			// TODO: will be better to use `t.factory` from 'typescript'.
			`declare module "${options.moduleEnvName}" {`,
			`export const env: ${typeDef}`,
			"}",
		],
		A.join("\n"),
		S.replaceAll("?: string | undefined", "?: string"),
	);

export const resolveId = (options: Options) =>
	F.ifElse(
		(id: string) => id === options.moduleEnvName,
		() => options.moduleEnvName,
		toUndefined,
	);

export const load = (env: Record<string, string>, options: Options) =>
	F.ifElse(
		(id: string) => id === options.moduleEnvName,
		() => Promise.resolve(createModuleEnv(env, options)),
		toUndefined,
	);

export const watchChange = (
	watchList: string[],
	onChange: (id: string) => void,
) =>
	F.ifElse(
		(id: string) => A.some(watchList, (x) => S.includes(path.basename(id), x)),
		onChange,
		toUndefined,
	);

export const build = (env: Record<string, string>, options: Options) =>
	options.schema.safeParseAsync(env).then(
		F.ifElse(
			({ success }) => success,
			(a) => Promise.resolve(a),
			(e) => Promise.reject(e?.error),
		),
	);

export { printTypeDefinition } from "./ast";
