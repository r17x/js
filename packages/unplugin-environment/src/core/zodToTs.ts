import { SyntaxKind, factory } from "typescript";
import { z } from "zod";
import { printTypeDefinition } from "./ast";

const primitives = {
	// string | number | symbol | bigint | boolean | null | undefined
	string: z.instanceof(z.ZodString),
	number: z.instanceof(z.ZodNumber),
	symbol: z.instanceof(z.ZodSymbol),
	bigint: z.instanceof(z.ZodBigInt),
	boolean: z.instanceof(z.ZodBoolean),
	null: z.instanceof(z.ZodNull),
	undefined: z.instanceof(z.ZodUndefined),

	// typescript keyword
	any: z.instanceof(z.ZodAny),
	unknown: z.instanceof(z.ZodUnknown),
	void: z.instanceof(z.ZodVoid),
	never: z.instanceof(z.ZodNever),
};

const _primitivesTSType = {
	string: SyntaxKind.StringKeyword,
	number: SyntaxKind.NumberKeyword,
	boolean: SyntaxKind.BooleanKeyword,
	null: SyntaxKind.NullKeyword,
	undefined: SyntaxKind.UndefinedKeyword,
	any: SyntaxKind.AnyKeyword,
	unknown: SyntaxKind.UnknownKeyword,
	void: SyntaxKind.VoidKeyword,
	never: SyntaxKind.NeverKeyword,
};

const stringLiteral = z
	.string()
	.transform((a) => factory.createStringLiteral(a));

const literalValue = z.union([
	stringLiteral,
	z.number().transform((a) => factory.createNumericLiteral(a)),
	// TODO: don't know how to handle this
	// z.symbol().transform(a => factory.createS),
	z
		.bigint()
		.transform((a) =>
			factory.createBigIntLiteral({
				negative: a.toString().startsWith("-"),
				base10Value: a.toString().replace("-", ""),
			}),
		),
	z
		.boolean()
		.transform((a) => (a ? factory.createTrue() : factory.createFalse())),
	z.null().transform(() => factory.createNull()),
	z
		.undefined()
		.transform(() =>
			factory.createKeywordTypeNode(SyntaxKind.UndefinedKeyword),
		),
]);

const literals = z
	.instanceof(z.ZodLiteral)
	.transform((v) => v.value)
	.pipe(
		z.union([
			z.string(),
			z.number(),
			z.symbol(),
			z.bigint(),
			z.boolean(),
			z.null(),
			z.undefined(),
		]),
	);

const zodEnumValue = (a: string[]) =>
	factory.createUnionTypeNode(
		a.map((b) => factory.createLiteralTypeNode(stringLiteral.parse(b))),
	);

const zodEnum = z
	.instanceof(z.ZodEnum)
	.transform((v) => v.options)
	.pipe(z.string().min(1).array());

const primitivesTransform = {
	string: factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
	number: factory.createKeywordTypeNode(SyntaxKind.NumberKeyword),
	boolean: factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword),
	null: factory.createKeywordTypeNode(SyntaxKind.NullKeyword),
	undefined: factory.createKeywordTypeNode(SyntaxKind.UndefinedKeyword),
	any: factory.createKeywordTypeNode(SyntaxKind.AnyKeyword),
	unknown: factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword),
	void: factory.createKeywordTypeNode(SyntaxKind.VoidKeyword),
	never: factory.createKeywordTypeNode(SyntaxKind.NeverKeyword),
};

const p = primitives.string
	.transform(() => primitivesTransform.string)
	.or(primitives.number.transform(() => primitivesTransform.number))
	.or(primitives.boolean.transform(() => primitivesTransform.boolean))
	.or(primitives.null.transform(() => primitivesTransform.null))
	.or(primitives.undefined.transform(() => primitivesTransform.undefined))
	.or(primitives.any.transform(() => primitivesTransform.any))
	.or(primitives.unknown.transform(() => primitivesTransform.unknown))
	.or(primitives.void.transform(() => primitivesTransform.void))
	.or(primitives.never.transform(() => primitivesTransform.never))
	.or(literals.transform((a) => literalValue.parse(a)))
	.or(zodEnum.transform((a) => zodEnumValue(a)));

export const toTSType = p.transform((a) => printTypeDefinition(a)).parse;
