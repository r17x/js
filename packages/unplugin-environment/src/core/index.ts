import path from "path";
import { fileURLToPath } from "url";
import { F } from "@mobily/ts-belt";
import { config as dotenvConfig } from "dotenv";
import type { UnpluginFactory } from "unplugin";
import * as Core from "./core";
import { log } from "./logger";
import type { PluginOption } from "./types";

export const unpluginFactory: UnpluginFactory<PluginOption> = (o) => {
	const options = Core.getOptions(o);
	const isProduction =
		["true", "1"].includes(process?.env?.CI || "") &&
		process?.env?.NODE_ENV !== "development";

	const env = process.env as Record<string, string>;
	/**
	 * NOTE:
	 * Vite not load env file automatically until user set configuration for env.
	 * So, we load all env file here (with `dotenv` package) and filter by our options.
	 */
	dotenvConfig({ processEnv: env });

	/**
	 * This writer function for generate `env.d.ts` file.
	 */
	const writer = () =>
		import("fs/promises").then(async (fs) => {
			fs.writeFile(
				path.resolve(path.dirname(fileURLToPath(import.meta.url)), "env.d.ts"),
				// it will be handle existing user configuration and client/server configuration.
				// for client/server configuration, we will generate on each and merged into one file `env.d.ts`.
				Core.mapOptions(options, {
					single: (o) => Core.createModuleDTS(env, o),
					clientServer: (o) =>
						[
							Core.createModuleDTS(env, o.client),
							"",
							Core.createModuleDTS(env, o.server),
						].join("\n"),
				}),
			);
		});

	const logError = (e: Error) => {
		log({
			kind: "error",
			message: e?.message || "something went wrong",
			prefix: "unplugin-environment",
		});
		isProduction && process.exit(1);
	};

	// use for resolve module by moduleEnvName.
	const resolveId = Core.mapOptions(options, {
		single: (o) => Core.resolveId(o),
		clientServer: (o) => (id: string) =>
			Core.resolveId(o.client)(id) || Core.resolveId(o.server)(id),
	});

	// use for load module by moduleEnvName.
	const loadInclude = Core.mapOptions(options, {
		single: (o) => Core.loadInclude(o),
		clientServer: (o) =>
			F.anyPass([Core.loadInclude(o.client), Core.loadInclude(o.server)]),
	});

	// use for load virtual module by moduleEnvName.
	const load = Core.load(env, options);

	// use for validation `schema` and `moduleEnvName` is use for generate `env.d.ts` file.
	const build = Core.mapOptions(options, {
		single: (o) => async () => {
			await Core.build(env, o).then(writer).catch(logError);
		},
		clientServer: (o) => async () => {
			await Promise.all([
				Core.build(env, o.client).then(writer),
				Core.build(env, o.server).then(writer),
			]).catch(logError);
		},
	});

	return {
		name: "unplugin-environment",
		enforce: "pre",
		resolveId,
		loadInclude,
		load,
		watchChange: Core.watchChange([".env"], () => {
			dotenvConfig({ processEnv: env });
		}),
		buildStart: build,
	};
};
