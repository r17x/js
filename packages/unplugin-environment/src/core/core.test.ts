import { beforeEach, describe, expect, it, vi } from "vitest";
import { z } from "zod";
import { zodToTs } from "zod-to-ts";
import * as Core from "./core";
import type { Options, PluginOption } from "./types";

describe("uplugin-environment:core", () => {
	beforeEach(() => {
		vi.clearAllMocks();
	});

	const casesSchema = [
		{
			schema: "",
			expected: "Error",
		},
		{
			schema: z.object({
				REACT_APP_NAME: z.string(),
			}),
			expected: z.ZodObject,
		},
		{
			schema: {
				REACT_APP_NAME: z.string(),
			},
			expected: z.ZodObject,
		},
	];

	it.each(casesSchema)(
		"$schema field of PluginOption is $expected",
		({ schema: s, expected }) => {
			const { data, success } = Core.schema.safeParse(s);
			if (expected === "Error") {
				expect(success).toBe(false);
			} else {
				expect(success).toBe(true);
				expect(data).instanceof(expected);
			}
		},
	);

	const casesOptions = [
		{
			options: "APP",
			expected: {
				match: "APP",
				schema: expect.any(z.ZodType),
				moduleEnvName: "@env",
			},
		},
		{
			options: ["APP", "!SERVER"],
			expected: {
				match: ["APP", "!SERVER"],
				schema: expect.any(z.ZodType),
				moduleEnvName: "@env",
			},
		},
		{
			options: {
				match: ["APP", "!SERVER"],
				schema: {
					APP_NAME: z.string(),
				},
			},
			expected: {
				match: ["APP", "!SERVER"],
				schema: expect.any(z.ZodType),
				moduleEnvName: "@env",
			},
		},
	];

	it.each(casesOptions)(
		"Valid $options as param of Environment",
		({ options, expected }) => {
			expect(Core.getOptions(options)).toStrictEqual(expected);
		},
	);

	it("valid merge zod schema", () => {
		expect(
			Core.printTypeDefinition(
				zodToTs(
					z
						.object({
							REACT_APP_NAME: z.string().optional(),
						})
						.merge(
							z.object({
								REACT_APP_VERSION: z.string(),
							}),
						),
				).node,
			),
		).matchSnapshot();
	});

	it("valid run build", () => {
		expect(
			Core.build(
				{ REACT_APP_NAME: "htmx_BTW" },
				Core.getOptions("REACT_APP") as Options,
			),
		).resolves.toStrictEqual({
			success: true,
			data: {
				REACT_APP_NAME: "htmx_BTW",
			},
		});
	});

	it("only run watchChange once when valid watchList", () => {
		const onChange = vi.fn();
		const watcher = Core.watchChange([".env"], onChange);
		watcher("/a/b/c/.env~~~~");
		watcher("/a/b/c/any");
		watcher("/a/b/c/any");
		watcher("/a/b/c/any");
		expect(onChange).toHaveBeenCalledTimes(1);
	});

	it("only for knowledge", () => {
		expect(
			Core.printTypeDefinition(
				zodToTs(z.record(z.any().optional(), z.string())).node,
			),
		).toMatchSnapshot();

		expect(
			Core.printTypeDefinition(
				zodToTs(
					z.object({
						REACT_APP_NAME: z.string(),
						REACT_APP_VERSION: z.string(),
						REACT_APP_AGE: z.number(),
					}),
				).node,
			),
		).toMatchSnapshot();
	});

	it("[DTS] valid create module type definition", () => {
		const options = Core.getOptions({
			match: "REACT_APP",
			schema: {
				REACT_APP_NAME: z.string(),
				REACT_APP_VERSION: z.string(),
				REACT_APP_AGE: z.number(),
				REACT_APP_PORT: z.coerce.number().min(1),
				REACT_APP_DATA: z
					.string()
					.transform((a) => JSON.parse(a))
					.pipe(
						z.object({
							age: z.number(),
							name: z.string(),
						}),
					),
			},
			moduleEnvName: "@myenv",
		});

		expect(
			Core.createModuleDTS(
				{
					REACT_APP_NAME: "htmx_BTW",
					REACT_APP_VERSION: "6.6.6",
					REACT_APP_DURATION: "123",
					REACT_APP_PORT: "8080",
					REACT_APP_DATA: JSON.stringify({ age: 123, name: "htmx_BTW" }),
					SECRET_KEY_TOKEN_: "1234567890",
				},
				options as Options,
			),
		).toMatchSnapshot();
	});

	it("valid create module env", () => {
		expect(() =>
			Core.load(
				{
					REACT_APP_NAME: "htmx_BTW",
					REACT_APP_VERSION: "6.6.6",
					REACT_APP_DURATION: "123",
					SECRET_KEY_TOKEN_: "1234567890",
				},
				Core.getOptions({
					match: "REACT_APP",
					schema: z.any(),
					moduleEnvName: "@myenv",
				}) as Options,
			)("@myenv"),
		).toThrowError();

		expect(
			Core.load(
				{
					REACT_APP_NAME: "htmx_BTW",
					REACT_APP_VERSION: "6.6.6",
					REACT_APP_DURATION: "123",
					SECRET_KEY_TOKEN_: "1234567890",
				},
				Core.getOptions({
					match: "REACT_APP",
					schema: z.object({
						REACT_APP_NAME: z.string(),
					}),
					moduleEnvName: "@myenv",
				}) as Options,
			)("@myenv"),
		).resolves.toMatchSnapshot();

		expect(
			Core.load(
				{
					REACT_APP_NAME: "htmx_BTW",
					REACT_APP_VERSION: "6.6.6",
					SECRET_KEY_TOKEN_: "1234567890",
				},
				Core.getOptions("REACT_APP") as Options,
			)("@env"),
		).resolves.toMatchSnapshot();
		expect(
			Core.createModuleEnv(
				{
					REACT_APP_NAME: "htmx_BTW",
					REACT_APP_VERSION: "6.6.6",
					SECRET_KEY_TOKEN_: "1234567890",
				},
				Core.getOptions("REACT_APP") as Options,
			),
		).toMatchSnapshot();
	});

	it("print type definition", () => {
		expect(
			Core.printTypeDefinition(
				zodToTs(
					z
						.object({
							REACT_APP_NAME: z.string(),
						})
						.strict(),
				).node,
			),
		).toBe(
			`{
    REACT_APP_NAME: string;
}`,
		);
	});

	it("valid parse env", () => {
		const mockEnv = {
			REACT_APP_PORT: "8080",
			REACT_APP_NAME: "htmx_BTW",
			REACT_APP_VERSION: "6.6.6",
			SECRET_KEY_TOKEN_: "1234567890",
		};

		expect(
			Core.getEnv(mockEnv, Core.getOptions("REACT_APP") as Options),
		).toStrictEqual({
			REACT_APP_NAME: "htmx_BTW",
			REACT_APP_VERSION: "6.6.6",
			REACT_APP_PORT: "8080",
		});

		expect(
			Core.getEnv(
				mockEnv,
				Core.getOptions(["REACT_APP", "!REACT_APP_VERSION"]) as Options,
			),
		).toStrictEqual({
			REACT_APP_NAME: "htmx_BTW",
			REACT_APP_PORT: "8080",
		});

		expect(
			Core.getEnv(
				mockEnv,
				Core.getOptions({
					match: ["REACT_APP", "!REACT_APP_VERSION"],
					schema: z.object({
						REACT_APP_NAME: z.string(),
						REACT_APP_PORT: z.coerce.number().min(1),
					}),
				}) as Options,
			),
		).toStrictEqual({
			REACT_APP_NAME: "htmx_BTW",
			REACT_APP_PORT: 8080,
		});
	});

	it("valid resolveId", () => {
		const options = Core.getOptions("REACT_APP");
		Core.mapOptions(options, {
			single: (o) => {
				expect(Core.resolveId(o)("@env")).toBe("@env");
				expect(Core.resolveId(o)("@env.ts")).toBeNull();
				expect(Core.resolveId(o)("@env.ts")).not.toBe("@env");
			},
			clientServer: () => {},
		});
	});

	it("valid options value", () => {
		expect(() => Core.getOptions("")).toThrowError();
		expect(() => Core.getOptions({} as PluginOption)).toThrowError();
		expect(() => Core.getOptions({ match: "" } as PluginOption)).toThrowError();
		expect(() =>
			Core.getOptions({ client: "", server: "" } as PluginOption),
		).toThrowError();

		expect(Core.getOptions("REACT_APP")).toStrictEqual({
			match: "REACT_APP",
			schema: expect.any(z.ZodType),
			moduleEnvName: "@env",
		});

		expect(Core.getOptions(["REACT_APP", "VITE_APP"])).toStrictEqual({
			match: ["REACT_APP", "VITE_APP"],
			schema: expect.any(z.ZodType),
			moduleEnvName: "@env",
		});

		expect(
			Core.getOptions({ client: "CLIENT_APP", server: "SERVER_APP" }),
		).toStrictEqual({
			client: {
				match: "CLIENT_APP",
				schema: expect.any(z.ZodType),
				moduleEnvName: "@env/client",
			},
			server: {
				match: "SERVER_APP",
				schema: expect.any(z.ZodType),
				moduleEnvName: "@env/server",
			},
		});

		expect(
			Core.getOptions({
				client: {
					match: "CLIENT_APP",
					schema: z.object({
						CLIENT_APP_NAME: z.string(),
					}),
					moduleEnvName: "@myenv/client",
				},
				server: {
					match: "SERVER_APP",
					schema: z.object({
						SERVER_APP_NAME: z.string(),
					}),
					moduleEnvName: "@myenv/server",
				},
			}),
		).toStrictEqual({
			client: {
				match: "CLIENT_APP",
				schema: expect.any(z.ZodType),
				moduleEnvName: "@myenv/client",
			},
			server: {
				match: "SERVER_APP",
				schema: expect.any(z.ZodType),
				moduleEnvName: "@myenv/server",
			},
		});
	});
});
