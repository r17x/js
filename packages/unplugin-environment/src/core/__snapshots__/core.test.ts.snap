// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`uplugin-environment:core > [DTS] valid create module type definition 1`] = `
"// DO NOT EDIT THIS FILE. IT IS GENERATED BY \`unplugin-environment@1.2.5\` - see https://npm.im/unplugin-environment

declare module "@myenv" {
export const env: {
    REACT_APP_NAME: string;
    REACT_APP_VERSION: string;
    REACT_APP_DURATION?: string;
    REACT_APP_PORT: number;
    REACT_APP_DATA: {
        age: number;
        name: string;
    };
    REACT_APP_AGE: number;
}
}"
`;

exports[`uplugin-environment:core > only for knowledge 1`] = `
"{
    [x: string]: string;
}"
`;

exports[`uplugin-environment:core > only for knowledge 2`] = `
"{
    REACT_APP_NAME: string;
    REACT_APP_VERSION: string;
    REACT_APP_AGE: number;
}"
`;

exports[`uplugin-environment:core > valid create module env 1`] = `
{
  "code": "export const env = {"REACT_APP_NAME":"htmx_BTW","REACT_APP_VERSION":"6.6.6"}",
}
`;

exports[`uplugin-environment:core > valid create module env 2`] = `
{
  "code": "export const env = {"REACT_APP_NAME":"htmx_BTW","REACT_APP_VERSION":"6.6.6","REACT_APP_DURATION":"123"}",
}
`;

exports[`uplugin-environment:core > valid create module env 3`] = `
{
  "code": "export const env = {"REACT_APP_NAME":"htmx_BTW","REACT_APP_VERSION":"6.6.6"}",
}
`;

exports[`uplugin-environment:core > valid merge zod schema 1`] = `
"{
    REACT_APP_NAME?: string | undefined;
    REACT_APP_VERSION: string;
}"
`;
